### PR TITLE
Add intake_xarray_kwargs to ThreddsCatalog

### LINKED ISSUE #51: add xarray_kwargs to ThreddsCatalog
I'm looking at GFS data e.g. https://thredds.ucar.edu/thredds/catalog/grib/NCEP/GFS/Global_0p25deg/catalog.html?dataset=grib/NCEP/GFS/Global_0p25deg/Best

you can't get it through pydap as
```
import xarray as xr
ds = xr.open_dataset("https://thredds.ucar.edu/thredds/dodsC/grib/NCEP/GFS/Global_0p25deg/Best", engine="pydap")
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 69191: ordinal not in range(128)
```
https://github.com/pydap/pydap/issues/196
I may create an issue in xarray to workout how to set https://github.com/pydap/pydap/issues/196#issuecomment-606301206 in the above call maybe using a backeng_kwargs.

Therefore, you have to get it through netcdf. e.g
```
import xarray as xr
ds = xr.open_dataset("https://thredds.ucar.edu/thredds/dodsC/grib/NCEP/GFS/Global_0p25deg/Best", engine="netcdf4")
ds = xr.open_dataset("https://thredds.ucar.edu/thredds/dodsC/grib/NCEP/GFS/Global_0p25deg/GFS_Global_0p25deg_20210913_1800.grib2", engine="netcdf4")

```

with netcdf as the driver here it's required to have a HTTPServer e.g. https://thredds.ucar.edu/thredds/catalog/grib/NCEP/GFS/Global_0p25deg/GFS_Global_0p25deg_20210913_1800.grib2/catalog.html?dataset=grib/NCEP/GFS/Global_0p25deg/GFS_Global_0p25deg_20210913_1800.grib2

Using this with `ThreddsCatalog` will pass it though to intake_xarray.netcdf.NetCDFSource.

However, intake_xarray.netcdf.NetCDFSource can't guess the engine so it would be nice to specify that in the ThreddsCatalog stage.

```
cat_url = "https://thredds.ucar.edu/thredds/catalog/grib/NCEP/GFS/Global_0p25deg/GFS_Global_0p25deg_20210913_1800.grib2/catalog.xml"

import intake
catalog = intake.open_thredds_cat(cat_url, driver="netcdf")
# nice to do intake.open_thredds_cat(cat_url, driver="netcdf", xarray_kwargs=dict(engine="netcdf4"))
source = catalog["GFS_Global_0p25deg_20210913_1800.grib2"]
ds = source().to_dask()
ValueError                                Traceback (most recent call last)
/var/folders/rf/26llfhwd68x7cftb1z3h000w0000gp/T/ipykernel_837/3050318223.py in <module>
----> 1 ds = source().to_dask()
      2 ds

~/miniconda3/envs/main/lib/python3.9/site-packages/intake_xarray/base.py in to_dask(self)
     67     def to_dask(self):
     68         """Return xarray object where variables are dask arrays"""
---> 69         return self.read_chunked()
     70 
     71     def close(self):

~/miniconda3/envs/main/lib/python3.9/site-packages/intake_xarray/base.py in read_chunked(self)
     42     def read_chunked(self):
     43         """Return xarray object (which will have chunks)"""
---> 44         self._load_metadata()
     45         return self._ds
     46 

~/miniconda3/envs/main/lib/python3.9/site-packages/intake/source/base.py in _load_metadata(self)
    234         """load metadata only if needed"""
    235         if self._schema is None:
--> 236             self._schema = self._get_schema()
    237             self.dtype = self._schema.dtype
    238             self.shape = self._schema.shape

~/miniconda3/envs/main/lib/python3.9/site-packages/intake_xarray/base.py in _get_schema(self)
     16 
     17         if self._ds is None:
---> 18             self._open_dataset()
     19 
     20             metadata = {

~/miniconda3/envs/main/lib/python3.9/site-packages/intake_xarray/netcdf.py in _open_dataset(self)
     90             url = fsspec.open(self.urlpath, **self.storage_options).open()
     91 
---> 92         self._ds = _open_dataset(url, chunks=self.chunks, **kwargs)
     93 
     94     def _add_path_to_ds(self, ds):

~/miniconda3/envs/main/lib/python3.9/site-packages/xarray/backends/api.py in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, backend_kwargs, *args, **kwargs)
    479 
    480     if engine is None:
--> 481         engine = plugins.guess_engine(filename_or_obj)
    482 
    483     backend = plugins.get_backend(engine)

~/miniconda3/envs/main/lib/python3.9/site-packages/xarray/backends/plugins.py in guess_engine(store_spec)
    146         )
    147 
--> 148     raise ValueError(error_msg)
    149 
    150 

ValueError: did not find a match in any of xarray's currently installed IO backends ['netcdf4', 'h5netcdf', 'scipy', 'cfgrib', 'pydap', 'rasterio', 'zarr']. Consider explicitly selecting one of the installed engines via the ``engine`` parameter, or installing additional IO dependencies, see:
http://xarray.pydata.org/en/stable/getting-started-guide/installing.html
http://xarray.pydata.org/en/stable/user-guide/io.html
```

### PR DESCRIPTION
**Pull request by andersy005**
- Fixes #51

### CODE REVIEW

**Comment by andersy005** [comment]
@raybellwaves, this is my attempt at addressing #51. I opted for `intake_xarray_kwargs` instead of `xarray_kwargs` so as to use it as a catch-all argument for all options that can be passed to the `intake_xarray` sources (such as https://github.com/intake/intake-xarray/blob/f1ca02d5c7734bb9de79074fe128ac5e7d598165/intake_xarray/netcdf.py#L48). 


Right now it appears that things are broken due to how fsspec is interfering with the `path` passed to xarray's `open_dataset`. 

```python
In [1]: import intake

In [2]: cat_url = "https://thredds.ucar.edu/thredds/catalog/grib/NCEP/GFS/Global_0p25deg/GFS_Global_0p25d
   ...: eg_20210913_1800.grib2/catalog.xml"

In [3]: catalog = intake.open_thredds_cat(cat_url, driver="netcdf", intake_xarray_kwargs={'xarray_kwargs'
   ...: : {'engine': "netcdf4"}})

In [4]: catalog = intake.open_thredds_cat(cat_url, driver="netcdf", intake_xarray_kwargs={'xarray_kwargs': {'engine': "netcdf4"}})

In [5]: source = catalog["GFS_Global_0p25deg_20210913_1800.grib2"]
```


```python
In [7]: source.to_dask()
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-75a159db6bd7> in <module>
----> 1 source.to_dask()

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/intake_xarray/base.py in to_dask(self)
     67     def to_dask(self):
     68         """Return xarray object where variables are dask arrays"""
---> 69         return self.read_chunked()
     70 
     71     def close(self):

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/intake_xarray/base.py in read_chunked(self)
     42     def read_chunked(self):
     43         """Return xarray object (which will have chunks)"""
---> 44         self._load_metadata()
     45         return self._ds
     46 

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/intake/source/base.py in _load_metadata(self)
    234         """load metadata only if needed"""
    235         if self._schema is None:
--> 236             self._schema = self._get_schema()
    237             self.dtype = self._schema.dtype
    238             self.shape = self._schema.shape

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/intake_xarray/base.py in _get_schema(self)
     16 
     17         if self._ds is None:
---> 18             self._open_dataset()
     19 
     20             metadata = {

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/intake_xarray/netcdf.py in _open_dataset(self)
     90             url = fsspec.open(self.urlpath, **self.storage_options).open()
     91 
---> 92         self._ds = _open_dataset(url, chunks=self.chunks, **kwargs)
     93 
     94     def _add_path_to_ds(self, ds):

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/xarray/backends/api.py in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, backend_kwargs, *args, **kwargs)
    495 
    496     overwrite_encoded_chunks = kwargs.pop("overwrite_encoded_chunks", None)
--> 497     backend_ds = backend.open_dataset(
    498         filename_or_obj,
    499         drop_variables=drop_variables,

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/xarray/backends/netCDF4_.py in open_dataset(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, group, mode, format, clobber, diskless, persist, lock, autoclose)
    549 
    550         filename_or_obj = _normalize_path(filename_or_obj)
--> 551         store = NetCDF4DataStore.open(
    552             filename_or_obj,
    553             mode=mode,

~/.mambaforge/envs/intake-thredds-dev/lib/python3.9/site-packages/xarray/backends/netCDF4_.py in open(cls, filename, mode, format, group, clobber, diskless, persist, lock, lock_maker, autoclose)
    351 
    352         if not isinstance(filename, str):
--> 353             raise ValueError(
    354                 "can only read bytes or file-like objects "
    355                 "with engine='scipy' or 'h5netcdf'"

ValueError: can only read bytes or file-like objects with engine='scipy' or 'h5netcdf'
```

**Comment by aaronspring** [comment]
This PR seems good to go. Can we merge?

File: tests/test_cat.py [hunk 1]
```diff
@@ -70,3 +70,18 @@ def test_ThreddsCatalog_simplecache_fails_opendap(thredds_cat_url):
...
         intake.open_thredds_cat(f'simplecache::{thredds_cat_url}', driver='opendap')
+
+
+def test_ThreddsCatalog_intake_xarray_kwargs():
+    """Test that ThreddsCatalog allows intake_xarray kwargs."""
+    cat = intake.open_thredds_cat(
+        'https://psl.noaa.gov/thredds/catalog/Datasets/noaa.ersst/catalog.xml',
+        driver='netcdf',
+        intake_xarray_kwargs={
+            'xarray_kwargs': {'decode_times': False, 'engine': 'netcdf4'},
```
Comment (aaronspring) [inline tests/test_cat.py:81]:
```suggestion
            'xarray_kwargs': {'decode_times': False, 'engine': 'scipy'},
```

File: tests/test_cat.py [hunk 2]
```diff
@@ -70,3 +70,18 @@ def test_ThreddsCatalog_simplecache_fails_opendap(thredds_cat_url):
...
         intake.open_thredds_cat(f'simplecache::{thredds_cat_url}', driver='opendap')
+
+
+def test_ThreddsCatalog_intake_xarray_kwargs():
+    """Test that ThreddsCatalog allows intake_xarray kwargs."""
+    cat = intake.open_thredds_cat(
+        'https://psl.noaa.gov/thredds/catalog/Datasets/noaa.ersst/catalog.xml',
+        driver='netcdf',
+        intake_xarray_kwargs={
+            'xarray_kwargs': {'decode_times': False, 'engine': 'scipy'},
```
Comment (aaronspring) [inline tests/test_cat.py:81]:
```suggestion
            'xarray_kwargs': {'decode_times': False, 'engine': 'h5netcdf'},
```

**Comment by aaronspring** [comment]
Just fixed that one test. Can this PR be merged?

**Review by aaronspring — APPROVED** [review]

**Comment by andersy005** [comment]
Thank you, @aaronspring! Let's go ahead and merge this as is. If there is any issue, we can address it later.

**Pull request merged**